### ISSUE TITLE
Error flushing data: Index and count must refer to a location within the buffer.

### ISSUE THREAD

**Issue by brianharcourt** [Bilal-S/iis2tomcat#83]
I've recently begun having an issue with one of my IIS sites.
I've upgraded to latest and still have the issue; tho now there is a "BIN" directory being created under every site's directory even tho it's a global install.  I've played with all the settings I can imagine may affect it but nothing seems to change.  All other sites, running off the same connector and same tomcat instance seem to be working without issue.

The initial visit to the site renders the typical "Generic Connector Communication Error" at the top of the results but the site (most times) reponds with the correct HTML after this message.

* The configuration parameters unique to this IIS site:
```
<?xml version="1.0" encoding="UTF-8"?>
<configuration>
    <system.webServer>
        <rewrite>
            <rules>
                <rule name="NNA today redirect" enabled="true" stopProcessing="true">
                    <match url="(.*)" />
                    <conditions logicalGrouping="MatchAny">
                        <add input="{HTTP_HOST}" pattern=".*nna-today\.com" />
                        <add input="{HTTP_HOST}" pattern=".*nissanamericastoday\.com" />
                        <add input="{HTTP_HOST}" pattern=".*nissan-today\.com" />
                        <add input="{HTTP_HOST}" pattern=".*todaynissan\.com" />
                        <add input="{HTTP_HOST}" pattern=".*nissan\.lonebuffalo\.com" />
                    </conditions>
                    <action type="Redirect" url="https://nissan-newstoday.com/{R:1}" />
                </rule>
                <rule name="nissan-newstoday HTTPS" enabled="true" stopProcessing="true">
                    <match url="(.*)" />
                    <conditions>
                        <add input="{HTTPS}" pattern="off" />
                        <add input="{HTTP_HOST}" pattern=".*objects\.lonebuffalo\.com" negate="true" />
                    </conditions>
                    <action type="Redirect" url="https://nissan-newstoday.com/{R:1}" />
                </rule>
            </rules>
        </rewrite>
    </system.webServer>
</configuration>
```

* Event Viewer entry
```Error flushing data: Index and count must refer to a location within the buffer. Parameter name: bytes at System.Text.UTF8Encoding.GetString(Byte[] bytes, Int32 index, Int32 count) at BonCodeAJP13.BonCodeAJP13Packet.GetString(Byte[] data, String& value, Int32 pos) at BonCodeAJP13.TomcatPackets.TomcatSendHeaders.GetHeaders() at BonCodeIIS.BonCodeCallHandler.PrintFlush(BonCodeAJP13PacketCollection flushCollection)```

* BonCodeConfigList
```
BonCode Active Configuration: 
-----------------------------------------------
BONCODEAJP13_ADOBE_SUPPORT = False
BONCODEAJP13_PORT = 8009
BONCODEAJP13_SERVER = localhost
BONCODEAJP13_AUTOFLUSHDETECTION_TICKS = 0
BONCODEAJP13_AUTOFLUSHDETECTION_BYTES = 0
MAX_BONCODEAJP13_CONCURRENT_CONNECTIONS = 0
BONCODEAJP13_LOG_LEVEL = 3
BONCODEAJP13_LOG_DIR = C:\lucee\AJP13\logs
BONCODEAJP13_LOG_FILE = BonCodeAJP13Connection
BONCODEAJP13_ENABLE_REMOTE_MANAGER = True
BONCODEAJP13_AUTOCOMPRESS = False
BONCODEAJP13_TEXT_MARK = System.String[]
BONCODEAJP13_MANAGER_URLS = System.String[]
BONCODEAJP13_MANAGER_FLEXURLS = System.String[]
BONCODEAJP13_HEADER_SUPPORT = True
BonCodeAjp13_DocRoot = 
BonCodeAjp13_PhysicalFilePath = 
BONCODEAJP13_BLACKLIST_HEADERS = URL,SERVER_SOFTWARE,SERVER_NAME,SERVER_PROTOCOL
BONCODEAJP13_WHITELIST_HEADERS = 
BONCODEAJP13_FLUSH_TIMEOUT = 30
BONCODEAJP13_SERVER_READ_TIMEOUT = 120000
BONCODEAJP13_SERVER_WRITE_TIMEOUT = 0
BONCODEAJP13_FORCE_SECURE_SESSION = False
BONCODEAJP13_REMOTEADDR_FROM = 
BONCODEAJP13_ALLOW_EMTPY_HEADERS = False
BONCODEAJP13_PATHINFO_HEADER = xajp-path-info
BONCODEAJP13_ENABLE_HTTPSTATUSCODES = True
BONCODEAJP13_TOMCAT_DOWN_URL = 
BONCODEAJP13_TOMCAT_STREAM_ERRORMSG = 
BONCODEAJP13_TOMCAT_TCPCLIENT_ERRORMSG = 
BONCODEAJP13_PATH_PREFIX = 
MAX_BONCODEAJP13_PACKET_LENGTH = 65536
MAX_BONCODEAJP13_USERDATA_LENGTH = 65530
BONCODEAJP13_ENABLE_CLIENTFINGERPRINT = False
BONCODEAJP13_FINGERPRINTHEADERS = REMOTE_ADDR,HTTP_ACCEPT,HTTP_ACCEPT_CHARSET,HTTP_ACCEPT_ENCODING,HTTP_ACCEPT_LANGUAGE,HTTP_USER_AGENT,UA_CPU,REMOTE_HOST
BONCODEAJP13_SKIP_IISCUSTOMERRORS = False
BONCODEAJP13_LOG_IPFILTER = 
BONCODEAJP13_REQUEST_SECRET = 
BONCODEAJP13_FORCE_GC = False
BONCODE_MODCFML_SECRET = d46a645883ec58823ef3ed16caa5fce51e89f38f7e99a10f61da2f8e9ac88cb5
BONCODE_DOCROOT_OVERRIDE = 
```

* settings file:
```
<Settings>
<Server>localhost</Server>
<Port>8009</Port>
<LogLevel>3</LogLevel>
<LogDir>C:\lucee\AJP13\logs</LogDir>
<EnableRemoteAdmin>True</EnableRemoteAdmin>
<EnableHeaderDataSupport>True</EnableHeaderDataSupport>
<ForceSecureSession>False</ForceSecureSession>
<AllowEmptyHeaders>False</AllowEmptyHeaders>
<PacketSize>65536</PacketSize>
<ModCFMLSecret>d46a645883ec58823ef3ed16caa5fce51e89f38f7e99a10f61da2f8e9ac88cb5</ModCFMLSecret>
</Settings>
```

*tomcat server.xml connector entry:
```
   <Connector port="8009" protocol="AJP/1.3" 
               redirectPort="8443" 
               tomcatAuthentication="false" 
               maxThreads="1000" 
               keepAliveTimeout="-1" 
               connectionTimeout="121000" 
               packetSize="65536"/>
```

* BonCode log of erroring request:
```
2019-06-22 13:58:14 BonCode AJP Connenctor version 1.0.41
2019-06-22 13:58:14 New Connection 1 of 0 to tomcat: [::1]:8009 ID: 1 [T-72]
2019-06-22 13:58:14 BonCodeAJP13.ServerPackets.BonCodeAJP13ForwardRequest GET /index.cfm 1429 bytes
 < appl-physical-path : E:\LBI\clipsheet\clients\nissan\
 < cert-keysize : 256
 < cert-secretkeysize : 2048
 < cert-server-issuer : C=US, S=VA, L=Herndon, O=Network Solutions L.L.C., CN=Network Solutions DV Server CA 2
 < cert-server-subject : OU=Domain Control Validated, OU=nsProtect Secure Xpress, CN=nissan-newstoday.com
 < gateway-interface : CGI/1.1
 < https : on
 < https-keysize : 256
 < https-server-issuer : C=US, S=VA, L=Herndon, O=Network Solutions L.L.C., CN=Network Solutions DV Server CA 2
 < https-server-subject : OU=Domain Control Validated, OU=nsProtect Secure Xpress, CN=nissan-newstoday.com
 < local-addr : 127.0.0.1
 < path-info : /index.cfm
 < remote-port : 51296
 < script-name : /index.cfm
 < server-port : 443
 < server-port-secure : 1
 < Connection : Keep-Alive
 < Accept : text/html, application/xhtml+xml, */*
 < Accept-Encoding : gzip, deflate
 < Accept-Language : en-US
 < Cookie : LUCEE_ADMIN_LANG=en; cfid=b08bd436-b471-426e-9a34-b045411ef131; cftoken=0; LUCEE_ADMIN_LASTPAGE=server.cache; RESULTCOUNT=25; CF_SESSION_NISSAN_NEWSTODAY_COM1097_LV=1561225851515
 < Host : nissan-newstoday.com
 < User-Agent : Mozilla/5.0 (Windows NT 6.3; WOW64; Trident/7.0; rv:11.0) like Gecko
 < DNT : 1
2019-06-22 13:58:15 BonCodeAJP13.TomcatPackets.TomcatSendBodyChunk  8188 bytes
2019-06-22 13:58:15 BonCodeAJP13.TomcatPackets.TomcatSendBodyChunk  8188 bytes
2019-06-22 13:58:15 BonCodeAJP13.TomcatPackets.TomcatSendBodyChunk  8188 bytes
2019-06-22 13:58:15 BonCodeAJP13.TomcatPackets.TomcatSendBodyChunk  8188 bytes
2019-06-22 13:58:15 BonCodeAJP13.TomcatPackets.TomcatSendBodyChunk  8188 bytes
2019-06-22 13:58:15 BonCodeAJP13.TomcatPackets.TomcatSendBodyChunk  8188 bytes
2019-06-22 13:58:15 BonCodeAJP13.TomcatPackets.TomcatSendBodyChunk  6540 bytes
2019-06-22 13:58:15 BonCodeAJP13.TomcatPackets.TomcatSendBodyChunk  4 bytes
2019-06-22 13:58:15 BonCodeAJP13.TomcatPackets.TomcatEndResponse  2 bytes
```

* BonCode log of immediately following request which is successful:
```
2019-06-22 13:58:22 New Connection 2 of 0 to tomcat: [::1]:8009 ID: 2 [T-72]
2019-06-22 13:58:22 BonCodeAJP13.ServerPackets.BonCodeAJP13ForwardRequest GET /index.cfm 1429 bytes
 < appl-physical-path : E:\LBI\clipsheet\clients\nissan\
 < cert-keysize : 256
 < cert-secretkeysize : 2048
 < cert-server-issuer : C=US, S=VA, L=Herndon, O=Network Solutions L.L.C., CN=Network Solutions DV Server CA 2
 < cert-server-subject : OU=Domain Control Validated, OU=nsProtect Secure Xpress, CN=nissan-newstoday.com
 < gateway-interface : CGI/1.1
 < https : on
 < https-keysize : 256
 < https-server-issuer : C=US, S=VA, L=Herndon, O=Network Solutions L.L.C., CN=Network Solutions DV Server CA 2
 < https-server-subject : OU=Domain Control Validated, OU=nsProtect Secure Xpress, CN=nissan-newstoday.com
 < local-addr : 127.0.0.1
 < path-info : /index.cfm
 < remote-port : 51299
 < script-name : /index.cfm
 < server-port : 443
 < server-port-secure : 1
 < Connection : Keep-Alive
 < Accept : text/html, application/xhtml+xml, */*
 < Accept-Encoding : gzip, deflate
 < Accept-Language : en-US
 < Cookie : LUCEE_ADMIN_LANG=en; cfid=b08bd436-b471-426e-9a34-b045411ef131; cftoken=0; LUCEE_ADMIN_LASTPAGE=server.cache; RESULTCOUNT=25; CF_SESSION_NISSAN_NEWSTODAY_COM1097_LV=1561225851515
 < Host : nissan-newstoday.com
 < User-Agent : Mozilla/5.0 (Windows NT 6.3; WOW64; Trident/7.0; rv:11.0) like Gecko
 < DNT : 1
2019-06-22 13:58:22 BonCodeAJP13.TomcatPackets.TomcatSendHeaders 151 bytes
 > Set-Cookie : CF_SESSION_NISSAN_NEWSTODAY_COM1097_LV=1561226302523;Path=/;Expires=Sat, 22-Jun-2019 18:18:22 UTC
 > Content-Type : text/html;charset=utf-8
 > Content-Length : 55640
2019-06-22 13:58:22 BonCodeAJP13.TomcatPackets.TomcatSendBodyChunk  8188 bytes
2019-06-22 13:58:22 BonCodeAJP13.TomcatPackets.TomcatSendBodyChunk  8188 bytes
2019-06-22 13:58:22 BonCodeAJP13.TomcatPackets.TomcatSendBodyChunk  8188 bytes
2019-06-22 13:58:22 BonCodeAJP13.TomcatPackets.TomcatSendBodyChunk  8188 bytes
2019-06-22 13:58:22 BonCodeAJP13.TomcatPackets.TomcatSendBodyChunk  8188 bytes
2019-06-22 13:58:22 BonCodeAJP13.TomcatPackets.TomcatSendBodyChunk  8188 bytes
2019-06-22 13:58:22 BonCodeAJP13.TomcatPackets.TomcatSendBodyChunk  6540 bytes
2019-06-22 13:58:22 BonCodeAJP13.TomcatPackets.TomcatSendBodyChunk  4 bytes
2019-06-22 13:58:22 BonCodeAJP13.TomcatPackets.TomcatEndResponse  2 bytes
```

**Comment by brianharcourt** [comment]
2nd request formerly immediately successful now simply hanging: no timeout; 'Waiting for . . .' without end.

**Comment by brianharcourt** [comment]
Resolved by NOT storing session data in cookie by removing the following from Application.cfc
```
<CFSET THIS.sessionStorage = "cookie" />
```
This app now uses a default session storage (a database).

**Comment by Bilal-S** [comment]
@brianharcourt 
I am glad to see that you were able to resolve this. 
The Lucee session store to cookie breaks the protocol and sends malformed packets which leads to this error. I have added this to the troubleshooting section in the documentation.

Best,
Bilal

**Comment by brianharcourt** [comment]
Thanks Bilal.  Your tool has served me well.
I did not have time to dig into deeper causes but for future Googlers with similar issues:

Cookie session storage has worked well for years for me.   I did NOT change the version of Lucee between successfully operating and the error.  

But I did change a number of parameters, all of which changing back did not seem to alleviate the issue.  Given my panicked state and the urgency with which I needed repair I'm certain I did not isolate all the variables that may yield better info.   

My guess is that some combination of connector variables allowed the broken cookie protocol to work but without the correct connector variables the broken protocol could not work.

B.
